allennlp/data/instance.py

```diff
@@ -33,13 +34,17 @@ def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
         for field in self.fields.values():
             field.count_vocab_items(counter)
 
-    def index_fields(self, vocab: Vocabulary):
+    def index_fields(self, vocab: Vocabulary) -> None:
         """
-        Converts all ``UnindexedFields`` in this ``Instance`` to ``IndexedFields``, given the
-        ``Vocabulary``.  This `mutates` the current object, it does not return a new ``Instance``.
+        Indexes all fields in this ``Instance`` using the provided ``Vocabulary``.
+        This `mutates` the current object, it does not return a new ``Instance``.
+        A ``DataIterator`` will call this on each pass through a dataset; we use the ``indexed``
+        flag to make sure that indexing only happens once.
```

Comment:
You might add a warning somewhere (maybe here, maybe somewhere else; not really sure) that if for some reason your vocab changes during the course of your program and you try to re-index stuff, you might get weird, unpredictable behavior.  I think it's _really_ unlikely that someone might do this, but still, a warning somewhere might be a good idea.

Reply:
done